agent/lib_monolog.c

```diff
@@ -0,0 +1,387 @@
+  if (nr_php_is_zval_valid_string(level_name)) {
+    level_name_string = nr_strdup(Z_STRVAL_P(level_name));
+  } else {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected level_name be a valid string, got type %d",
+                     __func__, Z_TYPE_P(level_name));
+  }
+
+  nr_php_zval_free(&level_name);
+  nr_php_arg_release(&level);
+
+  return level_name_string;
+}
+
+/*
+ * Purpose : Convert $message argument of Monolog\Logger::addRecord to a string
+ *
+ * Params  : Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log message; caller must free
+ */
+
+static char* nr_monolog_get_message(NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* message = NULL;
+  zval* message_arg = NULL;
+
+  message_arg = nr_php_arg_get(2, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  if (!nr_php_is_zval_valid_string(message_arg)) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected $message be a valid string, got type %d",
+                     __func__, Z_TYPE_P(message_arg));
+    message = nr_strdup("");
+  } else {
+    message = nr_strdup(Z_STRVAL_P(message_arg));
+  }
+  nr_php_arg_release(&message_arg);
+
+  return message;
+}
+
+/*
+ * Purpose : Format key of $context array's element as string
+ *
+ * Params  : zend_hash_key
+ * *
+ * Returns : A new string representing zval; caller must free
+ *
+ */
+static char* nr_monolog_fmt_context_key(zend_hash_key* hash_key) {
+  char* key_str = NULL;
+  zval* key = nr_php_zval_alloc();
+  if (nr_php_zend_hash_key_is_string(hash_key)) {
+    nr_php_zval_str(key, nr_php_zend_hash_key_string_value(hash_key));
+    key_str = nr_formatf("%s", Z_STRVAL_P(key));
+  } else if (nr_php_zend_hash_key_is_numeric(hash_key)) {
+    ZVAL_LONG(key, (zend_long)nr_php_zend_hash_key_integer(hash_key));
+    key_str = nr_formatf("%ld", (long)Z_LVAL_P(key));
+  } else {
+    /*
+     * This is a warning because this really, really shouldn't ever happen.
+     */
+    nrl_warning(NRL_INSTRUMENT, "%s: unexpected key type", __func__);
+    key_str = nr_formatf("unsupported-key-type");
+  }
+  nr_php_zval_free(&key);
+  return key_str;
+}
+
+/*
+ * Purpose : Format value of $context array's  element as string
+ *
+ * Params  : zval value
+ * *
+ * Returns : A new string representing zval; caller must free
+ *
+ */
+static char* nr_monolog_fmt_context_value(zval* zv) {
+  char* val_str = NULL;
+  zval* zv_str = nr_php_zval_alloc();
+
+  ZVAL_DUP(zv_str, zv);
+  convert_to_string(zv_str);
+  val_str = nr_strdup(Z_STRVAL_P(zv_str));
+  nr_php_zval_free(&zv_str);
+
+  return val_str;
+}
+
+/*
+ * Purpose : Format an element of $context array as "key => value" string
+ *
+ * Params  : zval value, pointer to string buffer to store formatted output
+ * and hash key
+ *
+ * Side effect : string buffer is reallocated with each call.
+ *
+ * Returns : ZEND_HASH_APPLY_KEEP to keep iteration
+ *
+ */
+static int nr_monolog_fmt_context_item(zval* value,
```

Comment:
Global comment: for the zvals that are passed in that we know we don't want to modify, consider passing them in like: `const zval* value`

Reply:
@zsistla This has been addressed in 15a434214eb78fd96f6d723cba1c4308147c1a93. 

**Note**
It was not possible to use `const` for `nr_monolog_get_level_name` `logger` argument, because it is used in `nr_php_object_has_method` and `nr_php_call`. The later functions signature has first argument's type as `zval*`. Casting from `const zval*` down to `zval*`, when using `logger` in those function calls, was not an option, because of `-Werror=cast-qual` compile switch, which gives the following compilation error:
```
php-agent/agent/lib_monolog.c:77:33: error: cast discards ‘const’ qualifier from pointer target type [-Werror=cast-qual]
   77 |   if (!nr_php_object_has_method((zval*) logger, "getLevelName" TSRMLS_CC)) {
      |                                 ^
```